bpf:classifiers: add Overlay VXLAN/Geneve classifiers

smagnani96 opened the pull request:
With this patch, the datapath is now able to signal Overlay packets during Trace/Drop notification events.
This is useful for two main reasons (treated in separate upcoming PRs):

1. Hubble/Monitor do not need to rely on fixed ports to decode overlay packets, they simply check this flag;
2. Datapath can then send more bytes during such events in case of Overlay packets when not needed

These new flags are solely used for tracing events and they do not interfere with bpf metrics computation.
The VXLAN/GENEVE flags can be computed either:

1. via ctx->mark (SKB only): we do set MARK_MAGIC_OVERLAY on the egress path of a packet.
2. via parsing L2/L3/L4 headers: matching for UDP with TUNNEL_PORT.

Please do refer to commit messages for a detailed description and considerations of the introduced changes.

Fixes: #39032

```release-note
 Introduce traffic classifiers CLS_FLAG_{VXLAN,GENEVE} for tracing overlay traffic.
```

julianwiedmann commented on bpf/lib/classifiers.h:
```diff
@@ -16,26 +19,147 @@ enum {
  */
 static __always_inline cls_flags_t
-ctx_classify(const struct __ctx_buff *ctx __maybe_unused)
+ctx_classify(struct __ctx_buff *ctx)
```
Did you consider whether the caller can already pass the ethertype here?

For `send_trace_notify()` this would be just `ctx_get_protocol()`. But `send_trace_notify4()` and `send_trace_notify6()` could optimize and pass `ETH_P_IP*` straight away.

smagnani96 replied:
I believe I have to thank you again. What you're suggesting I had already tried without significant improvements. BUT it triggered me once again, so I decided to give it another try. While doing that, I realized that the +140k verified instructions actually comes from another line of code in trace notifications:

```c
	__u64 cap_len = min_t(__u64, monitor ? : TRACE_PAYLOAD_LEN,
			      ctx_len);
```

The problem here is that:
1. such value is always computed even when the trace notification is not emitted (checked few lines below)
2. such value relies on monitor. Given that in function such as `tail_handle_snat_fwd_ipv6` we do lookup/modify the monitor value, the verifier complexity unexpectedly increases. In addition to my overlay changes + `ipv6_addr_copy()` being called in trace notification, that causes the +140k insns.

I did create a local branch to try all these optimization on top of main + these overlay classifiers v4/v6/unknown, and that's the outcome in the verifier worst case.

![image](https://github.com/user-attachments/assets/195efbd8-e67f-4c86-9150-40bdec5b8f49)

I guess that's a win, but unfortunately I'm going to put these changes in this PR. I tried to test the optimizations standalone and it doesn't seem to actually bring benefits 😅

julianwiedmann replied:
Just out of curiosity - is that on-top of the recent [bump to LLVM 19](https://github.com/cilium/cilium/pull/39632)?

smagnani96 replied:
Mmm code is rebased on top of main, but I'm using `quay.io/lvh-images/complexity-test-ci:6.1-20250428.152446`  (a month old), which inside has:

```bash
root@complexity-test-6:/host# clang -v
clang version 18.1.8 (https://github.com/llvm/llvm-project.git 3b5b5c1ec4a3095ab096dd780e84d7ab81f3d7ff)
```

smagnani96 replied:
Just saw from `/ci-verifier` that there's still +100k insns...
Despite I was using `complexity-test-ci` image, I was missing the step to update llvm in the image.
Now with LLVM v19 (as in CI), the new changes I just pushed are not helpful at all... in particular, it doesn't like this commit https://github.com/cilium/cilium/pull/39650/commits/95a66f64d6a66dd8f5b7fb02801d0717d66c0dd7.
I'll try to get back to this whole process and find a good compromise.

julianwiedmann replied:
> Despite I was using `complexity-test-ci` image, I was missing the step to update llvm in the image.

ah right, the GH verifier workflow ignores the LLVM version in the LVH image - and replaces it with what's available from the cilium-builder image.

> Now with LLVM v19 (as in CI), the new changes I just pushed are not helpful at all... in particular, it doesn't like this commit https://github.com/cilium/cilium/commit/95a66f64d6a66dd8f5b7fb02801d0717d66c0dd7.
I'll try to get back to this whole process and find a good compromise.

:(

smagnani96 replied:
[verifier-complexity.zip](https://github.com/user-attachments/files/20481199/verifier-complexity.zip)
Complexity now looks way better. Most impacted program is always bpf_host. On average, \~+10k insns.
There's still an outlier with kernel v6.6 (complexity using 6.1 config) in bpf_host/1 and bpf_host/2 for `tail_handle_snat_fwd_ipv6`: previous complexity \~790k insns, after patch \~840k insns (\~+40k, not too bad though wrt before, still that function has a bit of pressure).

I've tried many enhancements, and the only thing to help seems to be moving flags computation post the IPv6 address copy ONLY in `_send_trace_notify6` (moving that also in `_send_trace_notify` and `_send_trace_notify4` seems to worsen complexity for other functions).

from:
```c
        flags = ctx_classify(ctx, bpf_htons(ETH_P_IPV6)) | CLS_FLAG_IPV6;
	msg = (typeof(msg)) {
		...
	};

	if (orig_addr)
		ipv6_addr_copy(&msg.orig_ip6, orig_addr);
```
to
```c
	msg = (typeof(msg)) {
              ...
	};

	if (orig_addr)
		ipv6_addr_copy(&msg.orig_ip6, orig_addr);

	msg.flags = ctx_classify(ctx, bpf_htons(ETH_P_IPV6)) | CLS_FLAG_IPV6;
```

With this change the impact would be ~15k insns rather than ~40k. I'm a bit against it, here's why:

1. code is less aligned with other trace helpers;
2. Right now classifiers don't need to be computed if the trace is not emitted, that's fine. But soon I'd like to move from using TRACE_REASON_ENCRYPT to using classifiers CLS_FLAG_{WIREGUARD,IPSEC}. In that case, these two new classifiers are required to be computed even if the trace is not emitted, given we use that info for `update_trace_metrics`. To land this change, I'll need to re-move the flag computation before the `msg = {}` initialization, and I feel that would bring me back to this.
3.  I do also feel like moving instructions around is helpful in some configs but not in others.

@julianwiedmann WDYT? 🥵

julianwiedmann replied:
How much complexity cost do you currently have in the egress path from looking at the skb->mark **and** the L4 port? Can we defer the "fallback" support, and for now **only** look at the `skb->mark` when in the egress path? (in combination with my suggestion to consider the `obs_point`, where available)

smagnani96 replied:
Given the new changes in the code after your suggestions, the egress path should have a maximum complexity increase of ~15k instructions verified, differently than before (we had ~70k).

julianwiedmann reviewed (commented):
Very cool! Right now this only gets used deep down in the trace/drop notification, correct? And the drop path is boring from a complexity point-of-view, because it lives in its own tailcall.

Could you motivate the distinction between VXLAN and GENEVE? Is it so that hubble can easily parse the optional headers of the GENEVE packet?

julianwiedmann commented on bpf/lib/classifiers.h:
```diff
@@ -16,26 +19,160 @@ enum {
+			default:
+				__throw_build_bug();
+			}
+#endif /* OVERLAY_CLASSIFIERS */
```
For the `_send_trace_notify()` case (which is probably more sensitive to complexity, as it's inline), you could optimize here a bit further based on the `obs_point`.

For instance when it `to-netdev`, we only need/want to look at the `dport`.

julianwiedmann replied:
yeah thinking about this more, I think we're missing an opportunity to optimize. I believe there's exactly **one** case where looking at the `skb->mark` makes sense (in `to-netdev`), can we reflect that in the code?

julianwiedmann replied:
(after discussing :point_up: a bit more with Simone, `to-wireguard` can also observe the MAGIC_OVERLAY mark. nice :))

smagnani96 replied:
I just pushed new changes based on your suggestions.
We significantly preserve verifier complexity wrt the previous implementation.
As you suggested:

1. looking at packet mark: only in to-wireguard and to-network
2. parsing the headers: only in from-wireguard, from-network. There's an additional edge case for VinE that is to-stack in bpf_host: before forwarding the packet for encryption, we perform a trace notification. The new changes account also for this.

julianwiedmann reviewed (commented):
:wave: had a first look, some comments in-line.

One structural suggestion - I'd split off all the `send_trace_notifcy()` refactors where you need to touch the whole tree, and push them into their own PR. They are innocent enough (a bit of renaming + macro magic), but have much higher conflict potential. So let's get those in first, and then do a second PR that only focuses on extending the inner mechanics of the classifier.

smagnani96 committed:
bpf:classifiers: refactor tests to prepare for new classifiers

With this commit, we prepare our test suite for classifiers for the
new values introduced in subsequent commits.
This allows us to write reusable tests for both IPv4/IPv6, while also
specifying subtests for each possible category of flags (CLS_FLAG_*).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:classifiers: accept L3 protocol in ctx_classify

In #39794, we enabled passing the carried L3 protocol when calling
`send_trace_notify`. Let's use this argument for `ctx_classify`.
This allows to set the CLS_FLAG_IPV6 whenever we detect an IPV6 packet,
and not only when having a layer 3 IPv6 packet. In addition, this is
useful in subsequent commits where we extend classifiers: the new logic
will save verifier complexity and packet processing time given we have
already the protocol computed. For some edge case in which the variable
protocol is left to zero and not being set to the current protocol
(ex. from drop notification or when being called after a failing validate_ethertype),
we can compute the protocol from `ctx_classify` itself.

Here we also initialize the proto variable in bpf_lxc.c to avoid
compilation errors due to variable not being initialized. This was not
needed before as it was not being used, therefore removed during compilation.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:classifiers: add overlay VXLAN/Geneve classifiers

This commits introduces two new classifiers CLS_FLAG_{VXLAN,GENEVE}.
This are used to signal from datapath that the current packet being
traced/dropped is using one of these two protocols as underlay.
This is not used for computing bpf metrics so far, but will be useful
in a subsequent PR for decoding encap'd layer from Hubble/Monitor.
Having different signals for VXLAN and Geneve traffic is actually useful
so that Hubble/Monitor does not need to check the Cilium configuration
to retrieve which tunnel protocol is in use. This way, the datapath
automatically marks every overlay notification, and Hubble can easily
chose between the next decoding layer to pick. In addition, we can start
sending more/less packet during notification events based on the packet
itself, if it is native or overlay.

With this patch, we introduce two possible ways of determining a packet
is overlay:

1. ctx->mark:   available only with SKB, we look whether the mark contains
                MARK_MAGIC_OVERLAY. That value is usually set on our egress
                path. This has the advantage that is fast, as it does not
                require to further inspect packet L2/L3/L4 headers.
2. raw headers: this method inspects the whole packet layers up to L4 UDP.
                By looking at protocol and known ports (TUNNEL_PORT in use),
                we can determine a packet is overlay.

While (2) can always work in every condition, (1) works only when the mark
is set. This means that in all codepaths where the mark is cleared it can
fail. It would be still backed up by logic in (2), given that it will
always run in case (1) was not informative enough.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: move trace_point definition

Simply moving the definition of available `enum trace_point` into a
different file to avoid circular dependencies in subsequent commits.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: add trace observation point unknown

This patch introduces a new `enum trace_point` TRACE_POINT_UNKNOWN.
This is not intended to be used for `send_trace_notify` notification events,
as it will be translated into a compile error. The new observation point
is for special cases, such as the one enabled in the next commit.
Given that drop notification (1) do not account for the observation point
and (2) are executed in custom tailcalls, we can use the new TRACE_POINT_UNKNOWN
in the next commit where we optimize classifiers based on the trace
observation point. The TRACE_POINT_UNKNOWN allows to run all the possible
classification logic, both via mark and packet parsing.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:classifiers: optimize overlay based on observation point

This commit patches `ctx_classify` to account for the current trace
observation point. This allows to preserve both (1) packet processing
performance in all the codepaths where such overlay flags would never
be computed, and (2) preserve verifier complexity with a minimal impact.

The codepaths where we can encounter and trace overlay packets are:

1. to-{wireguard,netdev}: ctx->mark set to MARK_MAGIC_OVERLAY
2. from-{netdev,wireguard}: no mark, need to check UDP + tunnel port.
3. to-stack: with IPSec enabled, we trace TO_STACK events when forwarding
             overlay packets to the stack for encryption (VinE).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
